src/python/grpcio_observability/grpc_observability/server_call_tracer.cc

```diff
@@ -0,0 +1,238 @@
+  // response.
+  char stats_buf_[kMaxServerStatsLen];
+};
+
+
+void PythonOpenCensusServerCallTracer::RecordReceivedInitialMetadata(
+    grpc_metadata_batch* recv_initial_metadata) {
+  ServerMetadataElements sml;
+  FilterInitialMetadata(recv_initial_metadata, &sml);
+  path_ = std::move(sml.path);
+  method_ = GetMethod(path_);
+  auto tracing_enabled = PythonOpenCensusTracingEnabled();
+  GenerateServerContext(
+      tracing_enabled ? sml.tracing_slice.as_string_view() : "",
+      absl::StrCat("Recv.", method_), &context_);
+  if (PythonOpenCensusStatsEnabled()) {
+    std::vector<Label> labels = context_.Labels();
+    labels.emplace_back(Label{kServerMethod, std::string(method_)});
+    RecordIntMetric(kRpcServerStartedRpcsMeasureName, 1, labels);
+  }
+}
+
+
+void PythonOpenCensusServerCallTracer::RecordSendTrailingMetadata(
+    grpc_metadata_batch* send_trailing_metadata) {
+  // We need to record the time when the trailing metadata was sent to
+  // mark the completeness of the request.
+  elapsed_time_ = absl::Now() - start_time_;
+  if (PythonOpenCensusStatsEnabled() && send_trailing_metadata != nullptr) {
+    size_t len = ServerStatsSerialize(absl::ToInt64Nanoseconds(elapsed_time_),
+                                      stats_buf_, kMaxServerStatsLen);
+    if (len > 0) {
+      send_trailing_metadata->Set(
+          grpc_core::GrpcServerStatsBinMetadata(),
+          grpc_core::Slice::FromCopiedBuffer(stats_buf_, len));
+    }
+  }
+}
+
+
+void PythonOpenCensusServerCallTracer::RecordEnd(
+    const grpc_call_final_info* final_info) {
+  if (PythonOpenCensusStatsEnabled()) {
+    const uint64_t request_size = GetOutgoingDataSize(final_info);
+    const uint64_t response_size = GetIncomingDataSize(final_info);
+    double elapsed_time_ms = absl::ToDoubleMilliseconds(elapsed_time_);
+    std::vector<Label> labels = context_.Labels();
+    labels.emplace_back(Label{kServerMethod, std::string(method_)});
+    labels.emplace_back(Label{kServerStatus, std::string(StatusCodeToString(final_info->final_status))});
+    RecordDoubleMetric(kRpcServerSentBytesPerRpcMeasureName, static_cast<double>(response_size), labels);
+    RecordDoubleMetric(kRpcServerReceivedBytesPerRpcMeasureName, static_cast<double>(request_size), labels);
+    RecordDoubleMetric(kRpcServerServerLatencyMeasureName, elapsed_time_ms, labels);
+    RecordIntMetric(kRpcServerSentMessagesPerRpcMeasureName, sent_message_count_, labels);
+    RecordIntMetric(kRpcServerReceivedMessagesPerRpcMeasureName, recv_message_count_, labels);
+  }
+  if (PythonOpenCensusTracingEnabled()) {
+    context_.EndSpan();
+    if (IsSampled()) {
+      RecordSpan(context_.Span().ToCensusData());
+    }
+  }
+}
+
+//
+// PythonOpenCensusServerCallTracerFactory
+//
+
+grpc_core::ServerCallTracer*
+PythonOpenCensusServerCallTracerFactory::CreateNewServerCallTracer(
+    grpc_core::Arena* arena) {
+  return arena->ManagedNew<PythonOpenCensusServerCallTracer>();
```

Comment:
Discussed with @yashykt offline. To avoid this issue, we'll ignore the provided arena, call `new` / `delete` ourselves, and destroy the object at the end of the `RecordEnd` method.